bin/propolis-server/src/lib/vm/state_driver.rs

```diff
@@ -253,87 +253,58 @@ pub(super) async fn run_state_driver(
     let migration_in_id =
         ensure_request.migrate.as_ref().map(|req| req.migration_id);
 
-    let input_queue = Arc::new(InputQueue::new(
-        log.new(slog::o!("component" => "request_queue")),
-        match &ensure_request.migrate {
-            Some(_) => InstanceAutoStart::Yes,
-            None => InstanceAutoStart::No,
-        },
-    ));
-
-    let BuildVmOutput { vm_objects, migration_in } =
-        match super::startup::build_vm(
-            &log,
-            &vm,
-            &ensure_request,
-            &ensure_options,
-            &input_queue,
-            &mut state_publisher,
-        )
-        .await
-        {
-            Ok(objects) => objects,
-            Err((e, objects)) => {
-                state_publisher.update(ExternalStateUpdate::Instance(
-                    InstanceState::Failed,
-                ));
+    let ensure = VmEnsureNotStarted::new(
+        &log,
+        &vm,
+        &ensure_request,
+        &ensure_options,
+        ensure_result_tx,
+        &mut state_publisher,
+    );
+
+    let activated_vm =
+        if let Some(migrate_request) = ensure_request.migrate.as_ref() {
+            let migration = match crate::migrate::destination::initiate(
+                &log,
+                migrate_request,
+                ensure_options.local_server_addr,
+            )
+            .await
+            {
+                Ok(mig) => mig,
+                Err(e) => {
+                    ensure.fail(e.into()).await;
+                    return StateDriverOutput {
+                        state_publisher,
+                        final_state: InstanceState::Failed,
+                    };
+                }
+            };
 
-                vm.vm_init_failed(objects.is_some()).await;
-                let _ = ensure_result_tx
-                    .send(Err(VmError::InitializationFailed(e)));
+            match migration.run(ensure).await {
+                Ok(activated) => activated,
+                Err(_) => {
```

Comment:
what happens to the error returned by the migration attempt here? is it already recorded/logged someplace, or is it just getting thrown away here?